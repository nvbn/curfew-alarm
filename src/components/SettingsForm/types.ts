import { Future } from "../../utils/future";
import { Time } from "../../utils/time";

type TypedItemType = "time" | "number" | "action" | "boolean";

type BaseTypedItemProps<T, Q extends TypedItemType> = {
  id: string;
  title: string;
  type: Q;
  value: Future<T>;
  onChange: (newValue: T) => void;
};

export type TimeItemProps = BaseTypedItemProps<Time, "time">;

export type NumberItemProps = BaseTypedItemProps<number, "number">;

export type ActionItemProps = BaseTypedItemProps<boolean, "action">;

export type BooleanItemProps = BaseTypedItemProps<boolean, "boolean">;

export type TypedItemProps =
  | TimeItemProps
  | NumberItemProps
  | ActionItemProps
  | BooleanItemProps;
