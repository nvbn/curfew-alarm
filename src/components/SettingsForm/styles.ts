import { StyleSheet } from "react-native";

export default StyleSheet.create({
  item: {
    paddingHorizontal: 10,
    paddingVertical: 20,
    marginVertical: 8,
    marginHorizontal: 16,
    borderBottomWidth: 1,
    borderBottomColor: "#fafafa",
  },
  title: {
    fontSize: 24,
    paddingBottom: 10,
  },
  input: {
    fontSize: 18,
  },
  booleanItem: {
    alignItems: "flex-start",
  },
});
