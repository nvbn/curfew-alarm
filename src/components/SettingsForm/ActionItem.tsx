import React from "react";
import { Text, TouchableOpacity } from "react-native";

import { isReady } from "../../utils/future";
import styles from "./styles";
import { ActionItemProps } from "./types";

/**
 * Settings action that only visible when an action hasn't been performed yet.
 */
const ActionItem = ({
  id,
  title,
  value,
  onChange,
}: ActionItemProps): JSX.Element | null => {
  if (value || !isReady(value)) {
    return null;
  }

  return (
    <TouchableOpacity
      testID={`action-item-touchable-${id}`}
      style={styles.item}
      onPress={() => onChange(true)}
    >
      <Text style={styles.title}>{title}</Text>
    </TouchableOpacity>
  );
};

export default ActionItem;
