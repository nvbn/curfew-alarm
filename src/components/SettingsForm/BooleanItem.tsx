import React from "react";
import { Switch, Text, TouchableWithoutFeedback, View } from "react-native";

import { isReady } from "../../utils/future";
import styles from "./styles";
import { BooleanItemProps } from "./types";

/**
 * Checkbox/switch settings item for boolean values.
 */
const BooleanItem = ({
  id,
  title,
  value,
  onChange,
}: BooleanItemProps): JSX.Element => (
  <TouchableWithoutFeedback
    testID={`boolean-item-touchable-${id}`}
    onPress={() => onChange(!value)}
  >
    <View style={[styles.item, styles.booleanItem]}>
      <Text style={styles.title}>{title}</Text>
      {isReady(value) ? (
        <Switch
          value={value}
          onValueChange={onChange}
          testID={`boolean-item-switch-${id}`}
        />
      ) : (
        <Switch disabled />
      )}
    </View>
  </TouchableWithoutFeedback>
);

export default BooleanItem;
