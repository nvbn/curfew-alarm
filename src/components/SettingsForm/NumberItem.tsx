import React, { useCallback, useRef } from "react";
import { Text, TextInput, TouchableWithoutFeedback, View } from "react-native";

import { isReady } from "../../utils/future";
import styles from "./styles";
import { NumberItemProps } from "./types";

/**
 * Allows to modify numeric setting value.
 */
const NumberItem = ({
  id,
  title,
  value,
  onChange,
}: NumberItemProps): JSX.Element => {
  const onChangeAsNumber = useCallback(
    (textValue) => {
      const nextValue = parseInt(textValue, 10);
      if (!isNaN(nextValue)) {
        onChange(nextValue);
      }
    },
    [onChange],
  );

  const inputRef = useRef<TextInput>(null);

  return (
    <TouchableWithoutFeedback
      testID={`number-item-touchable-${id}`}
      onPress={() => inputRef.current?.focus()}
    >
      <View style={styles.item}>
        <Text style={styles.title}>{title}</Text>
        {isReady(value) ? (
          <TextInput
            ref={inputRef}
            testID={`number-item-input-${id}`}
            keyboardType="number-pad"
            defaultValue={value.toString()}
            onChangeText={onChangeAsNumber}
            style={styles.input}
          />
        ) : (
          <TextInput editable={false} />
        )}
      </View>
    </TouchableWithoutFeedback>
  );
};

export default NumberItem;
