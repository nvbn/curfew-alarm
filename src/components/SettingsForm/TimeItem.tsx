import RNDateTimePicker, {
  Event,
} from "@react-native-community/datetimepicker";
import React, { useCallback, useState } from "react";
import { Text, TouchableOpacity, View } from "react-native";

import { PLATFORM_OS_ANDROID, PlatformOS } from "../../dependencies/IPlatform";
import { Future, futureMap, isReady } from "../../utils/future";
import { dateToTime, formatTime, Time, timeToDate } from "../../utils/time";
import styles from "./styles";
import { TimeItemProps } from "./types";

type TimeItemIOSProps = {
  id: string;
  title: string;
  valueAsDate: Future<Date>;
  onChange: (event: Event, newValue?: Date) => void;
};

const TimeItemIOS = ({
  id,
  title,
  valueAsDate,
  onChange,
}: TimeItemIOSProps): JSX.Element => (
  <View style={styles.item}>
    <Text style={styles.title}>{title}</Text>
    {isReady(valueAsDate) && (
      <RNDateTimePicker
        testID={`time-item-ios-time-picker-${id}`}
        mode="time"
        value={valueAsDate}
        onChange={onChange}
      />
    )}
  </View>
);

type TimeItemAndroidProps = {
  id: string;
  title: string;
  value: Future<Time>;
  valueAsDate: Future<Date>;
  onChange: (event: Event, newValue?: Date) => void;
};

const TimeItemAndroid = ({
  id,
  title,
  value,
  valueAsDate,
  onChange,
}: TimeItemAndroidProps): JSX.Element => {
  const [showPicker, setShowPicker] = useState(false);

  const onChangeWithDismiss = useCallback(
    (event: Event, newValue?: Date) => {
      setShowPicker(false);
      onChange(event, newValue);
    },
    [setShowPicker, onChange],
  );

  return (
    <TouchableOpacity
      testID={`time-item-android-touchable-${id}`}
      style={styles.item}
      onPress={() => {
        setShowPicker(true);
      }}
    >
      <Text style={styles.title}>{title}</Text>
      {isReady(value) && isReady(valueAsDate) && (
        <>
          <Text style={styles.input}>{formatTime(value)}</Text>
          {showPicker && (
            <RNDateTimePicker
              testID={`time-item-android-time-picker-${id}`}
              mode="time"
              value={valueAsDate}
              onChange={onChangeWithDismiss}
            />
          )}
        </>
      )}
    </TouchableOpacity>
  );
};

type Props = TimeItemProps & {
  os: PlatformOS;
};

/**
 * Allows to modify settings value that store time.
 */
const TimeItem = ({ id, title, value, onChange, os }: Props): JSX.Element => {
  const valueAsDate = futureMap(value, timeToDate);

  const onChangeAsTime = useCallback(
    (_: Event, newValue?: Date): void =>
      newValue && onChange(dateToTime(newValue)),
    [onChange],
  );

  return os === PLATFORM_OS_ANDROID ? (
    <TimeItemAndroid
      id={id}
      title={title}
      value={value}
      valueAsDate={valueAsDate}
      onChange={onChangeAsTime}
    />
  ) : (
    <TimeItemIOS
      id={id}
      title={title}
      valueAsDate={valueAsDate}
      onChange={onChangeAsTime}
    />
  );
};

export default TimeItem;
