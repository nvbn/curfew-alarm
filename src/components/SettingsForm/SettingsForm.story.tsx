import { storiesOf } from "@storybook/react-native";
import React from "react";
import { Platform } from "react-native";

import { formatTime, Time } from "../../utils/time";
import SettingsForm from "./SettingsForm";

storiesOf("<SettingsForm />", module)
  .add("boolean true and time", () => (
    <SettingsForm
      options={[
        {
          id: "boolean",
          title: "boolean",
          type: "boolean",
          value: true,
          onChange: (enabled: boolean) => alert(`boolean is ${enabled}`),
        },
        {
          id: "time",
          title: "time",
          type: "time",
          value: { hour: 20, minute: 45 },
          onChange: (time: Time) => alert(`time is ${formatTime(time)}`),
        },
      ]}
      os={Platform.OS}
    />
  ))
  .add("boolean false and time", () => (
    <SettingsForm
      options={[
        {
          id: "boolean",
          title: "boolean",
          type: "boolean",
          value: false,
          onChange: (enabled: boolean) => alert(`boolean is ${enabled}`),
        },
        {
          id: "time",
          title: "time",
          type: "time",
          value: { hour: 20, minute: 45 },
          onChange: (time: Time) => alert(`time is ${formatTime(time)}`),
        },
      ]}
      os={Platform.OS}
    />
  ));
