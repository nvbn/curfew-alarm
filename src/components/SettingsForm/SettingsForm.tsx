import React from "react";
import { FlatList } from "react-native";

import BooleanItem from "./BooleanItem";
import NumberItem from "./NumberItem";
import TimeItem from "./TimeItem";
import {
  BooleanItemProps,
  NumberItemProps,
  TimeItemProps,
  TypedItemProps,
} from "./types";

type Props = {
  options: Omit<TypedItemProps, "os">[];
  os: string;
};

/**
 * A form with editable settings options.
 */
const SettingsForm = ({ options, os }: Props): JSX.Element => (
  <FlatList
    data={options}
    renderItem={({ item }) => {
      switch (item.type) {
        case "time":
          return <TimeItem {...(item as TimeItemProps)} os={os} />;
        case "number":
          return <NumberItem {...(item as NumberItemProps)} />;
        case "boolean":
          return <BooleanItem {...(item as BooleanItemProps)} />;
      }
    }}
    keyExtractor={(item) => item.id}
  />
);

export default SettingsForm;
