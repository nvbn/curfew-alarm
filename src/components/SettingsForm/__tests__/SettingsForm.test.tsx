/* eslint-disable @typescript-eslint/no-empty-function */
import { render } from "@testing-library/react-native";
import React from "react";

import {
  PLATFORM_OS_ANDROID,
  PLATFORM_OS_IOS,
} from "../../../dependencies/IPlatform";
import SettingsForm from "../SettingsForm";
import { TypedItemProps } from "../types";

describe("<SettingsForm />", () => {
  for (const os of [PLATFORM_OS_ANDROID, PLATFORM_OS_IOS]) {
    test(`can be rendered on ${os}`, () => {
      const options: TypedItemProps[] = [
        {
          id: "curfew-start",
          title: "Curfew begins at:",
          type: "time",
          value: { hour: 25, minute: 17 },
          onChange: () => {},
        },
        {
          id: "go-home",
          title: "Go home minutes before:",
          type: "number",
          value: 24,
          onChange: () => {},
        },
      ];

      const rendered = render(<SettingsForm options={options} os={os} />);

      expect(rendered.toJSON()).toMatchSnapshot();
    });
  }
});
