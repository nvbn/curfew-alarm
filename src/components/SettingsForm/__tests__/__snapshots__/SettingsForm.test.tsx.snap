// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SettingsForm /> can be rendered on android 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "id": "curfew-start",
        "onChange": [Function],
        "title": "Curfew begins at:",
        "type": "time",
        "value": Object {
          "hour": 25,
          "minute": 17,
        },
      },
      Object {
        "id": "go-home",
        "onChange": [Function],
        "title": "Go home minutes before:",
        "type": "number",
        "value": 24,
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "borderBottomColor": "#fafafa",
            "borderBottomWidth": 1,
            "marginHorizontal": 16,
            "marginVertical": 8,
            "opacity": 1,
            "paddingHorizontal": 10,
            "paddingVertical": 20,
          }
        }
        testID="time-item-android-touchable-curfew-start"
      >
        <Text
          style={
            Object {
              "fontSize": 24,
              "paddingBottom": 10,
            }
          }
        >
          Curfew begins at:
        </Text>
        <Text
          style={
            Object {
              "fontSize": 18,
            }
          }
        >
          25:17
        </Text>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "borderBottomColor": "#fafafa",
            "borderBottomWidth": 1,
            "marginHorizontal": 16,
            "marginVertical": 8,
            "paddingHorizontal": 10,
            "paddingVertical": 20,
          }
        }
        testID="number-item-touchable-go-home"
      >
        <Text
          style={
            Object {
              "fontSize": 24,
              "paddingBottom": 10,
            }
          }
        >
          Go home minutes before:
        </Text>
        <TextInput
          allowFontScaling={true}
          defaultValue="24"
          keyboardType="number-pad"
          onChangeText={[Function]}
          rejectResponderTermination={true}
          style={
            Object {
              "fontSize": 18,
            }
          }
          testID="number-item-input-go-home"
          underlineColorAndroid="transparent"
        />
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`<SettingsForm /> can be rendered on ios 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "id": "curfew-start",
        "onChange": [Function],
        "title": "Curfew begins at:",
        "type": "time",
        "value": Object {
          "hour": 25,
          "minute": 17,
        },
      },
      Object {
        "id": "go-home",
        "onChange": [Function],
        "title": "Go home minutes before:",
        "type": "number",
        "value": 24,
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "borderBottomColor": "#fafafa",
            "borderBottomWidth": 1,
            "marginHorizontal": 16,
            "marginVertical": 8,
            "paddingHorizontal": 10,
            "paddingVertical": 20,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 24,
              "paddingBottom": 10,
            }
          }
        >
          Curfew begins at:
        </Text>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "borderBottomColor": "#fafafa",
            "borderBottomWidth": 1,
            "marginHorizontal": 16,
            "marginVertical": 8,
            "paddingHorizontal": 10,
            "paddingVertical": 20,
          }
        }
        testID="number-item-touchable-go-home"
      >
        <Text
          style={
            Object {
              "fontSize": 24,
              "paddingBottom": 10,
            }
          }
        >
          Go home minutes before:
        </Text>
        <TextInput
          allowFontScaling={true}
          defaultValue="24"
          keyboardType="number-pad"
          onChangeText={[Function]}
          rejectResponderTermination={true}
          style={
            Object {
              "fontSize": 18,
            }
          }
          testID="number-item-input-go-home"
          underlineColorAndroid="transparent"
        />
      </View>
    </View>
  </View>
</RCTScrollView>
`;
